src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -308,36 +302,48 @@ internal SynchronizationContext ApplicableJobSyncContext
         {
             get
             {
-                using (this.JoinableTaskContext.NoMessagePumpSynchronizationContext.Apply())
+                if (this.JoinableTaskContext.IsOnMainThread)
                 {
-                    lock (this.JoinableTaskContext.SyncContextLock)
+                    if (this.mainThreadJobSyncContext == null)
```

Comment:
📝 This optimization is safe outside of the case where `ApplicableJobSyncContext` is serving as a barrier primitive (seems unlikely for this case)